alf/batch_norm.py

```diff
@@ -0,0 +1,330 @@
+            if torch.any(current_step >= self._max_steps):
+                warning_once("current_step should be smaller than "
+                             "max_steps. Got %s. Will be clamped to %s" %
+                             (current_step.max(), self._max_steps - 1))
+                current_step = current_step.clamp(max=self._max_steps - 1)
+                self._clamped = True
+        self._current_step = current_step
+
+    def reset_parameters(self):
+        """Reset the parameters."""
+        if self._track_running_stats:
+            for i in range(self._max_steps):
+                self._running_means[i].zero_()
+                self._running_vars[i].fill_(1)
+                self._num_batches_tracked[i].zero_()
+        if self._affine:
+            nn.init.ones_(self._weight)
+            nn.init.zeros_(self._bias)
+
+    def forward(self, input: torch.Tensor):
+        self._check_input_dim(input)
+
+        if self.training or not self._track_running_stats:
+            if self._track_running_stats:
+                current_step = self._current_step
+                if isinstance(current_step,
+                              torch.Tensor) and current_step.ndim != 0:
+                    assert torch.all(current_step == current_step[0]), (
+                        "all current_steps must be same for training.")
+                    current_step = current_step[0]
+                current_step = int(current_step)
+                running_mean = self._running_means[current_step]
+                running_var = self._running_vars[current_step]
+                if not self._clamped:
+                    num_batches_tracked = self._num_batches_tracked[
+                        current_step]
+                    num_batches_tracked.add_(1)
+                    if self._momentum is None:  # use cumulative moving average
+                        exponential_average_factor = 1.0 / float(
+                            num_batches_tracked)
+                    else:  # use exponential moving average
+                        exponential_average_factor = self._momentum
+                else:
+                    exponential_average_factor = 0.0
+            else:
+                running_mean = None
+                running_var = None
+                exponential_average_factor = 0.0
+            return F.batch_norm(
+                input,
+                running_mean,
+                running_var,
+                self._weight,
+                self._bias,
+                # whether the mini-batch stats should be used for normalization
+                # rather than the running stats.
+                # If current_step is out of limit, we will use the running stats
+                # for max_steps - 1 to normalize the batch so that we can keep
+                # training and eval consistent.
+                not self._clamped,
+                exponential_average_factor,
+                self._eps)
+        else:
```

Comment:
Document that this `else` stands for "Not during training and tracking running stats"?

Reply:
added.